common/asyncworkflow/queue/consumer/default_consumer.go

```diff
@@ -0,0 +1,228 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package consumer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/uber/cadence/.gen/go/sqlblobs"
+	"github.com/uber/cadence/client/frontend"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/codec"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/messaging"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+)
+
+const (
+	defaultShutdownTimeout = 5 * time.Second
+	defaultConcurrency     = 100
+)
+
+type defaultConsumer struct {
+	queueID         string
+	innerConsumer   messaging.Consumer
+	logger          log.Logger
+	scope           metrics.Scope
+	frontendClient  frontend.Client
+	ctx             context.Context
+	cancelFn        context.CancelFunc
+	wg              sync.WaitGroup
+	shutdownTimeout time.Duration
+	msgDecoder      codec.BinaryEncoder
+	concurrency     int
+}
+
+type Option func(*defaultConsumer)
+
+func WithConcurrency(concurrency int) Option {
+	return func(c *defaultConsumer) {
+		c.concurrency = concurrency
+	}
+}
+
+func New(
+	queueID string,
+	innerConsumer messaging.Consumer,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	frontendClient frontend.Client,
+	options ...Option,
+) *defaultConsumer {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	c := &defaultConsumer{
+		queueID:         queueID,
+		innerConsumer:   innerConsumer,
+		logger:          logger.WithTags(tag.Dynamic("queue-id", queueID)),
+		scope:           metricsClient.Scope(metrics.AsyncWorkflowConsumerScope),
+		frontendClient:  frontendClient,
+		ctx:             ctx,
+		cancelFn:        cancelFn,
+		shutdownTimeout: defaultShutdownTimeout,
+		msgDecoder:      codec.NewThriftRWEncoder(),
+		concurrency:     defaultConcurrency,
+	}
+
+	for _, opt := range options {
+		opt(c)
```

Comment:
this is an nice pattern

Reply:
yes I like this because it avoids infinitely growing constructor parameters which don't apply for some callers. also makes testing easier by providing overrides.